### PR TITLE
[core] Introduce executeFilter to TableRead

### PR DESCRIPTION
**Pull request by JingsongLi**
### Purpose

Computing engines such as Trino need to convert a row into Trino's own data structure after reading it, which is costly. We can actually perform filtering before returning to Trino to reduce the amount of data that needs to be converted.

Trino Connector can just use `TableRead.executeFilter()` to enable this feature.

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/predicate/CompoundPredicate.java [hunk 1]
```diff
@@ -52,6 +53,11 @@ public boolean test(Object[] values) {
         return function.test(values, children);
     }
 
+    @Override
+    public boolean test(InternalRow row) {
+        return false;
```
Comment (Aitozi) [inline paimon-common/src/main/java/org/apache/paimon/predicate/CompoundPredicate.java:58]: this should be `funciton.test(row, children)` ?

Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/predicate/CompoundPredicate.java:58]: you are right

File: paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java [hunk 2]
```diff
@@ -117,6 +118,61 @@ public void testBatchFilter() throws Exception {
                                 "2|21|201|binary|varbinary|mapKey:mapVal|multiset"));
     }
 
+    @Test
+    public void testBatchFilterWithExecution() throws Exception {
```
Comment (yuzelin) [inline paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java:122]: Is it necessary to add a same test in `PrimaryKeyFileStoreTableTest`?

Comment (JingsongLi) [reply paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java:122]: I think it is the same, so I think it is OK to be just here.

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateProjectionConverter.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+public class PredicateProjectionConverter implements PredicateReplaceVisitor {
+
+    private final Map<Integer, Integer> reversed;
+
+    public PredicateProjectionConverter(int[] projection) {
+        this.reversed = new HashMap<>();
+        for (int i = 0; i < projection.length; i++) {
+            reversed.put(projection[i], i);
+        }
+    }
```
Comment (tsreaper) [inline paimon-common/src/main/java/org/apache/paimon/predicate/PredicateProjectionConverter.java:35]: Is it possible that `projection` contains the same integer? Consider a table with three columns `a`, `b` and `c`, if I query `SELECT a, c, c, b FROM T`, then `projection` might be `[0, 2, 2, 1]`?

File: paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java (hunk 2 above)
Comment (yuzelin) [reply paimon-core/src/test/java/org/apache/paimon/table/AppendOnlyFileStoreTableTest.java:122]: Alright.

**Review by yuzelin — APPROVED** [review]
+1

File: paimon-common/src/main/java/org/apache/paimon/predicate/PredicateProjectionConverter.java (hunk 3 above)
Comment (JingsongLi) [reply paimon-common/src/main/java/org/apache/paimon/predicate/PredicateProjectionConverter.java:35]: It is OK to just pick a last column reference.

**Pull request merged**